fix & improve Sampler export

ronso0 opened the pull request:
* before `.xml` was added to any specified export file even if it ended with `.xml`
* remember the import/export directory

Holzhaus commented on src/mixer/samplerbank.cpp:
```diff
@@ -10,8 +11,17 @@
+namespace {
+
+const ConfigKey kConfigkeyLastImportExportDirectory(
+        "[Samplers]", "LastImportExportDirectory");
```
For new COs we wanted to use snake_case, I think 2e should use that for new config keys as well.

ronso0 replied:
okay. what's "2e"?

ronso0 replied:
ahhh "**we** should use that for new config keys as well

All done, ready again.

Holzhaus commented on src/mixer/samplerbank.cpp:
```diff
@@ -10,8 +11,17 @@
+namespace {
+
+const ConfigKey kConfigkeyLastImportExportDirectory(
+        "[Samplers]", "lastImportExportDirectory");
```
```suggestion
        "[Samplers]", "last
_import_export_directory");
```
New COs and config keys should use snake case.

ronso0 replied:
snake_case, now I got it, sry

Ready.

Holzhaus commented on src/mixer/samplerbank.cpp:
```diff
@@ -117,14 +136,22 @@ void SamplerBank::slotLoadSamplerBank(double v) {
     }
 
+    // Update the import/export directory
+    QFileInfo fileName(samplerBankPath);
```
```suggestion
    const QFileInfo fileName(samplerBankPath);
```

Holzhaus commented on src/mixer/samplerbank.cpp:
```diff
@@ -117,14 +136,22 @@ void SamplerBank::slotLoadSamplerBank(double v) {
 
+    QString lastImportExportDirectory = m_pConfig->getValue(
+            kConfigkeyLastImportExportDirectory,
+            QStandardPaths::writableLocation(QStandardPaths::MusicLocation));
```
Not sure if MusicLocation is the best fit for XML files, but I guess that is debatable and I don't really have a better suggestion.

ronso0 replied:
neither did I, that's why I picked it.
Though it is probably a good starting point, compared to having to navigate from the home folder to some other dir.

ywwg replied:
this seems fine to me too

Holzhaus reviewed (commented):
If the file "foo.xml" already exists and I type "foo.xml" into the filename field of save file dialog (or if I just select the existing file), I'm asked if I really want to overwrite the existing file. If I just type "foo", the ".xml" suffix is appended and the file is overwritten without warning.

ronso0 commented:
> If the file "foo.xml" already exists and I type "foo.xml" into the filename field of save file dialog (or if I just select the existing file), I'm asked if I really want to overwrite the existing file. If I just type "foo", the ".xml" suffix is appended and the file is overwritten without warning.

I'm really annoyed by havin to work around this Qt bug (which was fixed for Windows btw but not for Linux).
So we could either put that dialog into a loop (while file exists) and feed it back as starting point so Qt's "Overwrite?" can kick in?
Or we show out own oerwrite dialog, like we do fr controller mappings?

ronso0 commented:
Actually I'm in favor of **A** (IINM we'd just need to feed back the fileFilter and the file location) even though that would create small UX hickup: when clicking Save that would open the dialog again with the file and the extension being preselected. Then Save and you'd get asked if to overwrite or not.

daschuer commented:
We already have the override Dialog in main? I think this is a good solution, because the is use case for overriding when the user has lets say 4 favorit sets and wants to adjust one of it. 

So my vote it **B** 

Sorry, If I have missed the point why this is a question here.

daschuer commented:
Oh, I now understand. We bypass the current solution by the added extension. 
**A** works for me as well.

ronso0 commented:
> **B** we show out own oerwrite dialog, like we do for controller mappings?
Downside of this is that it's needed for Linux only -- thus we'd need to make that dialog conditional to not have two confirm dialogs on Win/macOS which use the native file dialogs (which don't suffer the bug), or set `QFileDialog::DontUseNativeDialog` which feels wrong.

All in all this feels like re-inventing the wheel. We could as well use an own QFileDialog with the root issue fixed.

ronso0 commented:
I implemented **A** in #4531. It's the last two commits and for now only implemented for exporting playlists from the Playlist feature.
Please check if this works for you.

ronso0 committed:
sampler import/export: verify file extension, remember directory

Co-authored-by: Daniel Schürmann <daschuer@mixxx.org>

ronso0 commented:
This is now based on #4531. I'll rebase as soon as that is merged.

ronso0 commented on src/mixer/samplerbank.cpp:
```diff
@@ -33,28 +44,40 @@ SamplerBank::SamplerBank(PlayerManager* pPlayerManager)
 
+const QString SamplerBank::samplerFiletype() {
+    return tr("Mixxx Sampler Banks (*.xml)");
+}
```
Doesn't feel this is the smartest way to define a `tr` string which is used multple times.
Any other ways to do that?

ronso0 replied:
QT_TR_NOOP seems to be the weapon of choice, put in the anonymous namespace.
`const QString kSamplerFileType = QT_TR_NOOP("Mixxx Sampler Banks (*.xml)");`

Let me know what you think.

Swiftb0y replied:
Shouldn't `const QString kSamplerFileType = QObject::tr("Mixxx Sampler Banks (*.xml)"); work just as well? According to Qt docs, `QT_TR_NOOP` marks a string for "delayed translation", but I don't know what that means, maybe lazy loading of the translated string on first use?

ronso0 replied:
I'll try to test the translation for both to be sure. Never did that before though..

daschuer replied:
Does nothing: 
```#define QT_TR_NOOP(x) x```
https://github.com/qt/qtbase/blob/eba9196304dcc9129a7bb68db87c6fb42c953090/src/corelib/global/qglobal.h#L1356

But ... lupdate finds it so that later use of tr() with this variable can be translated. 

This is demontarted here: 
https://github.com/qt/qtbase/blob/fb33e2a8e82cfc28cf3c3c4bad657e0400640cb2/src/corelib/doc/snippets/code/src_corelib_global_qglobal.cpp#L389

daschuer replied:
This means the current code translates the string on every call. The proposed solution of @Swiftb0y only translates one time. However I am not sure if the translation is initialized before the string is initialized.

ronso0 replied:
Unfortunately I'm not familiar with the translation workflow, so I appreciate any advice how to proceed here.

daschuer commented:
#4531 is now merged. Can you rebase this again?

daschuer commented on src/mixer/samplerbank.cpp:
```diff
@@ -2,16 +2,28 @@
+
+const ConfigKey kConfigkeyLastImportExportDirectory(
+        "[Samplers]", "last_import_export_directory");
+const QString kSamplerFileType = QT_TR_NOOP("Mixxx Sampler Banks (*.xml)");
```
```suggestion
const char kSamplerFileType[] = QT_TR_NOOP("Mixxx Sampler Banks (*.xml)");
```

ronso0 replied:
Thanks!

Swiftb0y replied:
can explain why a `char[]` is preferred here? `QT_TR_NOOP` is used nowhere else in mixxx so explaining why its used here definitively deserves explanation in the code IMO.

ronso0 replied:
_ping_
Shall we add a comment, then merge?

Swiftb0y replied:
Yes please

ronso0 replied:
hehe, this was meant for @daschuer

daschuer replied:
QT_TR_NOOP should always used with char[] or char*, because the result will be used as key for fetching the translated string in tr() 

https://doc.qt.io/qt-5/qobject.html#tr

Using QString will convert from char[] and back to char*

daschuer commented on src/mixer/samplerbank.cpp:
```diff
@@ -38,23 +50,30 @@ void SamplerBank::slotSaveSamplerBank(double v) {
-            &fileFilter);
-    if (samplerBankPath.isNull() || samplerBankPath.isEmpty()) {
+            lastImportExportDirectory,
+            kSamplerFileType,
```
```suggestion
            tr(kSamplerFileType),
```

ronso0 replied:
...and every similiar occurence I assume.

ronso0 commented:
Thank you!

I rebased onto 2.3, added recent `tr` suggestions and made the dialog preselect `samplers.xml`.
Please check, and if it's okay I'll squash the commits.

daschuer reviewed (approved):
LGTM. 
There is still room for improving redundant translations lookups, but let's move on to more important tasks

ronso0 commented:
Alright, done here.

Holzhaus reviewed (approved):
LGTM, thank you!

daschuer commented on src/mixer/samplerbank.cpp:
```diff
@@ -2,16 +2,28 @@
+
+const ConfigKey kConfigkeyLastImportExportDirectory(
+        "[Samplers]", "last_import_export_directory");
+const char kSamplerFileType[] = QT_TR_NOOP("Mixxx Sampler Banks (*.xml)");
```
```suggestion
// This is used in multiple tr() calls below which accepts const char* as a key. 
// lupdate finds the single string here.   
const char kSamplerFileType[] = QT_TR_NOOP("Mixxx Sampler Banks (*.xml)");
```

ronso0 committed:
samplerbank: explain use of QT_TR_NOOP

Co-authored-by: Daniel Schürmann <daschuer@mixxx.org>

ronso0 commented:
Okay, great. Ready!

daschuer commented:
Oh sorry, I guess that is my dog food. Sorry I will prepare a fixed branch you can reset too, or we can merge this and fix it upstream.

ronso0 commented:
I don't spot the difference? I commited your comment suggestion, what else is missing?

ronso0 commented:
Done and CI all green.

daschuer commented:
Thank you!
